### PR TITLE
Adds code to build optimized regexes from arrays of strings

### PR DESCRIPTION
**Pull request by Sesquipedalian**
The `build_regex()` function creates optimized regular expressions from an array of strings.

An optimized regex built using this function will be much faster than a simple regex built using `implode('|', $strings)` --- anywhere from several times to several orders of magnitude faster.

However, the time required to build the optimized regex is approximately equal to the time it takes to execute the simple regex. Therefore, it is only worth calling this function if the resulting regex will be used more than once.

Because PHP places an upper limit on the allowed length of a regex, very large arrays may be split and returned as multiple regexes. In such cases, you will need to iterate through all elements of the returned array in order to test all possible matches.

### CODE REVIEW

**Comment by Sesquipedalian** [comment]
This PR is preparation for another that will take care of [a TODO item in parse_bbc()'s autolinker](https://github.com/SimpleMachines/SMF2.1/blob/release-2.1/Sources/Subs.php#L1876):

`Add a cron job to download the current list of TLDs from http://data.iana.org/TLD/tlds-alpha-by-domain.txt, process it into a regex, and store it somewhere. Then replace this variable with a call of some sort to retrieve the stored regex.`

I'm submitting in two separate PRs because the regex building functionality that this PR provides is a feature unto itself.

**Comment by jdarwood007** [comment]
I'm trying to understand this.  Got any usage case on how this works?  Seems like we could just build the regex manually elsewhere without introducing all of this.

**Comment by Sesquipedalian** [comment]
> Got any usage case on how this works?

Sure. :)

First, a bit of background:

The autolinker regex in parse_bbc() needs an explicit list of top level domains to check in order to match naked domain names (e.g. `example.com` in `I like to go to example.com for all my example needs`). Right now parse_bbc() is just using a hard-coded simple alternation regex for this job (see [here](https://github.com/SimpleMachines/SMF2.1/blob/release-2.1/Sources/Subs.php#L1877)). 

However, that regex only matches a small subset of the valid TLDs that exist now. The list of valid TLDs stands at a bit over 1500 entries as of today, and it will continue to grow. Keeping up with this ever growing list is not viable using a hard coded regex. If we want to keep up, we need a way to keep our list up to date programmatically.

So, I've put together some code that downloads the latest list of valid top level domain names from iana.org. Then I process that list into a regex and cache the result for later use in parse_bbc().

Now, the use case for build_regex():

Checking for over 1500 possibilities in a simple alternation regex would be pretty slow. But by feeding that list to build_regex, we can get a regex to check for all of them in a fraction of the time. By building this optimized regex once and then caching it for, say, a week, we can execute the autolinker in parse_bbc() much, much faster than we could using the simple alternation regex. Given how many times parse_bbc() is executed to put together even a single page for display, optimizing speed in this function is a big deal.

In developing this algorithm, I did a good bit of benchmarking. Searching in the same block of text using the same the list of TLDs as my array of keywords in both cases, here are the results I got when testing with 100 repetitions of each operation (compiling the optimized regex, executing the optimized regex, executing a simple alternation regex):

```
Compile time (100 repetitions): 6.4747741222382
Optimized execution time(100 repetitions): 0.2268660068512
Simple execution time(100 repetitions): 8.7822730541229
```

As you can see, the optimized regex executed 38.7 times faster than the simple regex with this set of strings. I also tested using variously sized collections of words taken from /usr/share/dict/words. With small sets of randomly selected words, the speed boost from optimization would be "only" 2-5 times faster. With really large sets of words (e.g. 4000) the boost was typically in the vicinity of 100 (!) times faster (though whether above or below varied a good bit from one specific set of strings to another).

So basically, whenever one has a list of alternative strings to search for using a regex, and one is going to search for that set of strings more than once, it makes sense to get build_regex() to build the regex for you.

> Seems like we could just build the regex manually elsewhere without introducing all of this.

Not really. Or at least, not if we want to end up with a regex anywhere nearly so efficient as what this produces.

To illustrate, here's the output from build_regex for the list of valid TLDs:

```
(?>சிங்கப்பூர்|فلسطين|ファッション|همراه|संगठन|বাংলা|భారత్|дети|تونس|شبكة|भारत|ভারত|ਭਾਰਤ|ભારત|ලංකා|クラウド|グーグル|ポイント|组织机构|電訊盈科|ком|укр|қаз|հայ|קום|قطر|كوم|कॉम|नेट|คอม|ไทย|みんな|ストア|セール|我爱你|淡马锡|诺基亚|飞利浦|ελ|ею|سو(?>دان|رية)|გე|コム|世界|企业|佛山|信息|健康|八卦|嘉里(?>大酒店|)|在线|娱乐|家電|工行|广东|微博|慈善|时尚|書籍|机构|游戏|澳門|点看|珠宝|移动|联通|谷歌|购物|通販|集团|食品|餐厅|삼성|한국|a(?>vianca|kdn|a(?>rp|a)|b(?>udhabi|ogado|arth|le|b(?>ott|vie|)|c)|c(?>ademy|c(?>ountant(?>s|)|enture)|o|t(?>ive|or)|)|d(?>ult|ac|s|)|e(?>tna|ro|g|)|f(?>amilycompany|l|)|g(?>akhan|ency|)|i(?>g(?>o|)|r(?>force|bus|tel)|)|l(?>faromeo|i(?>baba|pay)|l(?>finanz|state|y)|s(?>ace|tom)|)|m(?>sterdam|fam|ica|e(?>rican(?>express|family)|x)|)|n(?>alytics|droid|quan|z)|o|p(?>artments|p(?>le|))|q(?>uarelle|)|r(?>amco|chi|my|pa|t(?>e|)|)|s(?>sociates|da|ia|)|t(?>torney|hleta|)|u(?>ction|spost|di(?>ble|o|)|t(?>hor|o(?>s|))|)|w(?>s|)|x(?>a|)|z(?>ure|))|b(?>a(?>uhaus|yern|idu|by|n(?>a(?>narepublic|mex)|d|k)|r(?>efoot|gains|c(?>elona|lay(?>card|s))|)|)|b(?>va|c|t|)|c(?>g|n)|d|e(?>ntley|rlin|er|st(?>buy|)|a(?>uty|ts)|t|)|f|g|h(?>arti|)|i(?>ble|ke|ng(?>o|)|d|o|z|)|j|l(?>ue|a(?>nco|ck(?>friday|))|o(?>ckbuster|omberg|g))|m(?>s|w|)|n(?>pparibas|l|)|o(?>ehringer|utique|ats|fa|nd|m|o(?>ts|k(?>ing|)|)|s(?>tik|ch)|t|)|r(?>idgestone|adesco|ussels|o(?>adway|ther|ker)|)|s|t|u(?>dapest|siness|gatti|ild(?>ers|)|zz|y)|v|w|y|z(?>h|))|c(?>sc|a(?>fe|b|l(?>vinklein|l|)|m(?>era|p|)|n(?>cerresearch|on)|p(?>etown|ital(?>one|))|r(?>avan|tier|ds|e(?>er(?>s|)|)|s|)|s(?>ino|a|h)|t(?>ering|)|)|b(?>re|a|n|s)|c|d|e(?>nter|rn|b|o)|f(?>a|d|)|g|h(?>intai|urch|eap|loe|a(?>se|n(?>nel|el)|t)|r(?>istmas|ysler|ome)|)|i(?>priani|rcle|sco|t(?>adel|i(?>c|)|y(?>eats|))|)|k|l(?>eaning|aims|ub(?>med|)|i(?>ck|ni(?>que|c))|o(?>thing|ud)|)|m|n|o(?>rsica|ffee|ach|des|l(?>lege|ogne)|m(?>cast|sec|m(?>unity|bank)|p(?>uter|a(?>ny|re))|)|n(?>dos|s(?>truction|ulting)|t(?>ractors|act))|o(?>king(?>channel|)|l|p)|u(?>ntry|rses|pon(?>s|))|)|r(?>icket|uises|edit(?>union|card|)|own|s|)|u(?>isinella|)|v|w|x|y(?>mru|ou|)|z)|d(?>rive|clk|ds|hl|np|tv|a(?>bur|nce|d|t(?>ing|sun|e)|y)|e(?>mocrat|gree|al(?>er|s|)|nt(?>ist|al)|si(?>gn|)|l(?>ivery|oitte|ta|l)|v|)|i(?>amonds|gital|rect(?>ory|)|et|s(?>co(?>unt|ver)|h)|y)|j|k|m|o(?>wnload|mains|dge|ha|c(?>tor|s)|g|t|)|u(?>pont|rban|bai|ck|n(?>lop|s))|v(?>ag|r)|z)|e(?>quipment|ve(?>rbank|nts)|a(?>rth|t)|c(?>o|)|d(?>eka|u(?>cation|))|e|g|m(?>erck|ail)|n(?>terprises|gineer(?>ing|)|ergy)|p(?>ost|son)|r(?>icsson|ni|)|s(?>urance|tate|q|)|t|u(?>rovision|s|)|x(?>traspace|change|p(?>osed|ress|ert)))|f(?>tr|yi|a(?>mily|ge|rm(?>ers|)|i(?>rwinds|th|l)|n(?>s|)|s(?>hion|t))|e(?>edback|dex|rr(?>ari|ero))|i(?>at|lm|na(?>nc(?>ial|e)|l)|sh(?>ing|)|d(?>elity|o)|r(?>mdale|e(?>stone|))|t(?>ness|)|)|j|k|l(?>i(?>ghts|ckr|r)|o(?>rist|wers)|y)|m|o(?>undation|o(?>dnetwork|tball|)|r(?>sale|ex|um|d)|x|)|r(?>esenius|l|o(?>gans|nt(?>door|ier))|)|u(?>rniture|tbol|ji(?>xerox|tsu)|nd))|g(?>a(?>rden|me(?>s|)|l(?>l(?>ery|up|o)|)|p|)|b(?>iz|)|d(?>n|)|e(?>orge|nt(?>ing|)|a|)|f|g(?>ee|)|h|i(?>ft(?>s|)|v(?>ing|es)|)|l(?>ob(?>al|o)|a(?>de|ss)|e|)|m(?>ail|bh|o|x|)|n|o(?>daddy|l(?>d(?>point|)|f)|o(?>d(?>hands|year)|g(?>le|)|)|p|t|v)|p|q|r(?>een|ipe|oup|a(?>inger|phics|tis)|)|s|t|u(?>ardian|cci|ge|ru|i(?>tars|de)|)|w|y)|h(?>dfc(?>bank|)|gtv|sbc|bo|a(?>mburg|ngout|us)|e(?>alth(?>care|)|l(?>sinki|p)|r(?>mes|e))|i(?>samitsu|tachi|phop|v)|k(?>t|)|m|n|o(?>ckey|rse|use|me(?>depot|goods|s(?>ense|))|st(?>ing|)|l(?>dings|iday)|n(?>eywell|da)|t(?>eles|mail|)|w)|r|t(?>c|)|u(?>ghes|)|y(?>undai|att))|i(?>piranga|inet|kano|bm|fm|wc|c(?>bc|e|u)|d|e(?>ee|)|l|m(?>amat|db|mo(?>bilien|)|)|n(?>vestments|dustries|f(?>initi|o)|g|k|s(?>titute|ur(?>ance|e))|t(?>uit|e(?>rnational|l)|)|)|o|q|r(?>ish|)|s(?>elect|maili|t(?>anbul|)|)|t(?>au|v|))|j(?>nj|a(?>guar|va)|c(?>b|p)|e(?>welry|tzt|ep|)|l(?>c|l)|m(?>p|)|o(?>b(?>urg|s)|t|y|)|p(?>morgan|rs|)|u(?>niper|egos))|k(?>uokgroup|aufen|ddi|fh|e(?>rry(?>properties|logistics|hotels)|)|g|h|i(?>tchen|nd(?>er|le)|wi|a|m|)|m|n|o(?>matsu|sher|eln)|p(?>mg|n|)|r(?>ed|d|)|w|y(?>oto|)|z)|l(?>gbt|ds|pl(?>financial|)|a(?>dbrokes|caixa|salle|m(?>borghini|er)|n(?>xess|c(?>aster|ome|ia)|d(?>rover|))|t(?>robe|ino|)|w(?>yer|)|)|b|c|e(?>clerc|frak|ase|xus|g(?>al|o))|i(?>ghting|aison|lly|psy|xil|dl|fe(?>insurance|style|)|ke|m(?>ited|o)|n(?>coln|de|k)|v(?>ing|e)|)|k|o(?>ndon|an(?>s|)|ft|tt(?>e|o)|ve|c(?>ker|us)|l)|r|s|t(?>d(?>a|)|)|u(?>ndbeck|pin|x(?>ury|e)|)|v|y)|m(?>ba|a(?>serati|drid|keup|ttel|cys|i(?>son|f)|n(?>agement|go|)|r(?>shalls|riott|ket(?>ing|s|))|)|c(?>kinsey|d(?>onalds|)|)|d|e(?>lbourne|tlife|et|d(?>ia|)|m(?>orial|e)|n(?>u|)|o|)|g|h|i(?>crosoft|ami|l|n(?>i|t)|t(?>subishi|))|k|l(?>b|s|)|m(?>a|)|n|o(?>torcycles|scow|par|bi(?>ly|)|da|e|i|m|n(?>tblanc|ster|ash|ey)|r(?>tgage|mon)|v(?>i(?>star|e)|)|)|p|q|r|s(?>d|)|t(?>pc|n|r|)|u(?>seum|tu(?>elle|al)|)|v|w|x|y|z)|n(?>ba|hk|tt|yc|a(?>goya|dex|me|vy|b|t(?>ionwide|ura)|)|c|e(?>ustar|c|t(?>bank|flix|work|)|w(?>s|)|x(?>us|t(?>direct|))|)|f(?>l|)|g(?>o|)|i(?>nja|ssa(?>n|y)|co|k(?>on|e)|)|l|o(?>kia|rt(?>hwesternmutual|on)|w(?>ruz|tv|)|)|p|r(?>a|w|)|u|z)|o(?>kinawa|saka|pen|ff(?>ice|)|oo|vh|b(?>server|i)|l(?>dnavy|ayan(?>group|)|lo)|m(?>ega|)|n(?>yourside|e|g|l(?>ine|))|r(?>a(?>cle|nge)|g(?>anic|)|i(?>entexpress|gins))|t(?>suka|t))|p(?>ccw|ub|a(?>mperedchef|ssagens|ge|n(?>asonic|erai)|r(?>is|s|t(?>ners|s|y))|y|)|e(?>t|)|f(?>izer|)|g|h(?>armacy|ilips|ysio|oto(?>graphy|s|)|)|i(?>oneer|aget|zza|c(?>s|t(?>ures|et))|d|n(?>g|k|))|k|l(?>a(?>ce|y(?>station|))|u(?>mbing|s)|)|m|n(?>c|)|o(?>litie|ker|hl|rn|st)|r(?>ess|ime|a(?>merica|xi)|o(?>gressive|tection|pert(?>ies|y)|mo|d(?>uctions|)|f|)|u(?>dential|)|)|s|t|w(?>c|)|y)|q(?>pon|ue(?>bec|st)|vc|a)|r(?>yukyu|a(?>cing|dio|id)|e(?>cipes|xroth|view(?>s|)|hab|st(?>aurant|)|a(?>d|l(?>estate|t(?>or|y)))|d(?>umbrella|stone|)|i(?>se(?>n|)|t)|n(?>t(?>als|)|)|p(?>ublican|air|ort)|)|i(?>ghtathome|c(?>oh|h(?>ardli|))|o|p)|o(?>gers|deo|om|c(?>her|ks)|)|s(?>vp|)|u(?>hr|n|)|w(?>e|))|s(?>fr|a(?>arland|kura|fe(?>ty|)|ms(?>club|ung)|rl|ve|xo|l(?>on|e)|n(?>dvik(?>coromant|)|ofi)|p(?>o|)|s|)|b(?>i|s|)|c(?>johnson|ience|a|b|h(?>aeffler|midt|warz|ule|o(?>larships|ol))|o(?>r|t)|)|d|e(?>rvices|lect|cur(?>ity|e)|ner|ven|at|ek|s|w|x(?>y|)|)|g|h(?>riram|ell|a(?>ngrila|rp|w)|i(?>ksha|a)|o(?>uji|es|p(?>ping|)|w(?>time|))|)|i(?>lk|te|n(?>gles|a)|)|j|k(?>i(?>n|)|y(?>pe|)|)|l(?>ing|)|m(?>art|ile|)|n(?>cf|)|o(?>ft(?>bank|ware)|hu|c(?>cer|ial)|l(?>utions|ar)|n(?>g|y)|y|)|p(?>readbetting|iegel|ace|ot)|r(?>l|t|)|t(?>ream|yle|ud(?>io|y)|a(?>ples|da|r(?>hub|)|t(?>oil|e(?>bank|farm)))|c(?>group|)|o(?>ckholm|r(?>age|e))|)|u(?>zuki|cks|pp(?>ort|l(?>ies|y))|r(?>gery|f)|)|v|w(?>atch|i(?>ftcover|ss))|x|y(?>mantec|stems|dney|)|z)|t(?>a(?>ipei|obao|rget|lk|b|t(?>too|a(?>motors|r))|x(?>i|))|c(?>i|)|d(?>k|)|e(?>masek|nnis|am|ch(?>nology|)|va|l(?>e(?>fonica|city)|))|f|g|h(?>eat(?>er|re)|d|)|i(?>ckets|ffany|enda|aa|ps|r(?>es|ol))|j(?>maxx|x|)|k(?>maxx|)|l|m(?>all|)|n|o(?>shiba|day|kyo|ols|ray|tal|urs|wn|p|y(?>ota|s)|)|r(?>ust|a(?>ining|vel(?>channel|ers(?>insurance|)|)|d(?>ing|e))|v|)|t|u(?>nes|shu|be|i)|v(?>s|)|w|z)|u(?>connect|ol|ps|a|b(?>ank|s)|g|k|n(?>i(?>versity|com)|o)|s|y|z)|v(?>laanderen|a(?>cations|n(?>guard|a)|)|c|e(?>ntures|gas|r(?>mögensberat(?>ung|er)|sicherung|isign)|t|)|g|i(?>ajes|king|llas|rgin|deo|g|n|p|s(?>ion|ta(?>print|)|a)|v(?>a|o)|)|n|o(?>lkswagen|yage|dka|t(?>ing|e|o))|u(?>elos|))|w(?>hoswho|me|a(?>rman|tch(?>es|)|ng(?>gou|)|l(?>mart|ter|es))|e(?>ather(?>channel|)|b(?>site|cam|er)|d(?>ding|)|i(?>bo|r))|f|i(?>lliamhill|en|ki|n(?>dows|ners|e|))|o(?>lterskluwer|odside|r(?>ld|k(?>s|))|w)|s|t(?>c|f))|x(?>finity|peria|erox|box|xx|yz|i(?>huan|n))|y(?>un|a(?>maxun|chts|ndex|hoo)|e|o(?>dobashi|kohama|ga|u(?>tube|))|t)|z(?>uerich|ero|one|ip(?>po|)|a(?>ppos|ra|)|m|w)|б(?>ел|г)|м(?>кд|о(?>сква|н))|о(?>нлайн|рг)|р(?>ус|ф)|с(?>айт|рб)|ا(?>بوظبي|رامكو|مارات|یران|ل(?>سعودية|جزائر|عليان|اردن|مغرب))|ب(?>ازار|ھارت|يتك)|ع(?>راق|مان)|م(?>ليسيا|صر|و(?>بايلي|قع))|இ(?>ந்தியா|லங்கை)|中(?>文网|信|国|國)|公(?>司|益)|台(?>湾|灣)|商(?>城|店|标)|大(?>众汽车|拿)|手(?>机|表)|政(?>务|府)|新(?>加坡|闻)|网(?>址|店|站|络)|香(?>格里拉|港)|닷(?>넷|컴))
```

I wouldn't want to build that manually.

**Comment by jdarwood007** [comment]
Well honestly for 1500 TLDs and more coming, at what point do we declare it a lost cause to try and validate the tld and go with making sure it just makes sense.  I would be curious to see how other software is handling this as it doesn't seem like a winning battle as more TLDs are expected to come online.

We could also use filter_var with FILTER_VALIDATE_URL and not require schema, path and querystring (basically making it a domain match).

Or go for something like this

```
^(?!\-)(?:[a-zA-Z\d\-]{0,62}[a-zA-Z\d]\.){1,126}(?!\d+)[a-zA-Z\d]{1,63}$
```

As taken from: https://stackoverflow.com/questions/3026957/how-to-validate-a-domain-name-using-regex-php
Which basically is just seeing if it makes sense and falls into what is currently considered as a valid domain name structure.

**Comment by Sesquipedalian** [comment]
Well, that one wouldn't work at all with international domain names. Rather than `[a-zA-Z\d]` (or more succinctly, `\w`), one needs to use `[\p{L}\p{M}\p{N}]` if one want to match non-Latin characters. That's what we are using already in the autolinker regex in parse_bbc(). And in the case of TLDs specifically, we actually only need `\p{L}\p{M}`, since no existing TLDs include numbers (although that could change one day).

FILTER_VALIDATE_URL can only handle URIs, not IRIs, so we can't rely on it if we want to support languages not written in Latin script. I wish it did, because it would make life much easier. But it doesn't, so I had to come up with code to deal with that in #3573 (see the part that deals with #1389).

Merely checking to see if a string has some characters, a dot, and then some more characters is not at all reliable when trying to find domain names embedded in a string of text. Again, if FILTER_VALIDATE_URL worked for IRIs, we could make it work. But since it doesn't, a simple check like that would generate a terribly high number of false positives. Any time a user simply forgot to press space at the end of a sentence, the autolinker would turn that into a hyperlink.

Basically, dealing with naked domain names only works reliably if one uses a known list of TLDs. When dealing with fully formed URLs, or even partial URLs that have at least some more of the standard parts included (a scheme and a colon at the front, a `www`, a trailing slash and a path of some sort, etc.), it is possible to detect them based on the presence of those other telltale parts. But for naked domain names, the only way to distinguish them from strings of random characters that happen to have a dot in the middle someplace is by checking whether the part after the dot is a valid TLD.

**Comment by Sesquipedalian** [comment]
But at any rate, I'm not sure why you seem to be opposed to the inclusion of this function as an option. I mean, its not like anyone is forced to use it. It is another tool in the box, to be used when appropriate. My upcoming PR to deal with TLDs is one use case. There may be others we can think of once the tool is in the box.

**Comment by Sesquipedalian** [comment]
Another good use case would be in parsesmileys(). On [line 2648](https://github.com/SimpleMachines/SMF2.1/blob/release-2.1/Sources/Subs.php#L2648) of Subs.php, we implode a list of strings into a regex to search for all the possible smileys. Since almost all of the smileys start with the same character, using build_regex() could provide a sizeable speed boost compared to what we get using implode(). And since the final regex is stored in a static variable so that in can be reused every time the function is called (which is often), the speed boost would pay off many times over.

**Comment by jdarwood007** [comment]
Not opposed to it, just was trying to understand why it was useful, you gave a good reason.  Then trying to decide on the TLD issue if we are going in the correct direction with detection of a fairly complex regex rather than something more simple.

**Comment by Sesquipedalian** [comment]
Fair enough. Tone is often hard to gauge in written communication. :)

**Comment by MissAllSunday** [comment]
I'm in favor of doing this :) could the cron job can be used to create the optimized regex too?  that way we can store it/them on a $modSettings and only update it when the cron job runs again, say once a month or similar, I'm not too familiar with how ften the TLD list gets updated.  Or whenever the $modSetting is empty, whatever happens first

**Comment by Sesquipedalian** [comment]
That's pretty much the plan, yes. :) I was going to use the cache to store the compiled regex, but storing it in $modSettings would probably work better. Good idea, @MissAllSunday.

**Comment by Oldiesmann** [comment]
For the cron job, I assume that would use SMF's built-in automated task feature?

**Comment by Sesquipedalian** [comment]
Yes

**Comment by MissAllSunday** [comment]
Let me know if you want to merge this or do you want to work on it some more.

**Comment by Sesquipedalian** [comment]
Please merge this, yes.
